### PR TITLE
Report cache access information

### PR DESCRIPTION
**Pull request by jaebaek**
Based on the Vulkan spec, Vulkan application can get the cache hit/miss
information for the pipeline and all shader stages using
[Pipeline Creation Feedback](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/chap11.html#pipelines-creation-feedback).

Since LLPC uses/maintains the pipeline and the shader caches, the ICD
needs the help from LLPC to get the cache hit/miss information.
This commit lets LLPC report the information to the ICD i.e., XGL.

[This XGL PR](https://github.com/GPUOpen-Drivers/xgl/pull/68) reports
the cache hit/miss information through `VkPipelineCreationFeedbackCreateInfoEXT`
data structure based on the information given by LLPC.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Comment by jaebaek** [comment]
Related PR: https://github.com/GPUOpen-Drivers/xgl/pull/68

File: llpc/context/llpcCompiler.cpp
Comment (kuhar) [inline llpc/context/llpcCompiler.cpp:1586]: Why not define `stageCacheAccesses` as `std::array`? I'd also think that `any_of` should work with a plain C array of static size, no?

File: llpc/include/llpc.h
Comment (kuhar) [inline llpc/include/llpc.h:121]: remove this comment

File: llpc/include/llpc.h
Comment (kuhar) [inline llpc/include/llpc.h:129]: remove this comment

File: llpc/include/llpc.h
Comment (kuhar) [inline llpc/include/llpc.h:123]: I can't parse these two comments. Maybe something like `Indicates whether pipeline cache access was a hit`?

File: llpc/include/llpc.h
Comment (kuhar) [inline llpc/include/llpc.h:131]: same as above

**Review by kuhar — CHANGES_REQUESTED** [review]
I'm not sure if the public headers should be guarded by `LLPC_INTERFACE_MAJOR_VERSION`

File: llpc/include/llpc.h
Comment (s-perron) [reply llpc/include/llpc.h]: I agree.  You should output a more descriptive comment like:

```
bool pipelineCacheHit;                               // True if the pipeline was found in the cache.
StageCacheInfo stageCacheAccesses[ShaderStageCount]; // True if an unlinked shader was found in the cache for the stage.
```

File: llpc/context/llpcCompiler.cpp
Comment (s-perron) [inline llpc/context/llpcCompiler.cpp:781]: You probably want to use an "ArrayRef" for this paramenter, and everywhere else it is passed internally.

File: llpc/context/llpcCompiler.cpp
Comment (s-perron) [reply llpc/context/llpcCompiler.cpp]: The problem is that he wants to leave out the element referring to the compute stage.  I don't think this is worth asserting. The program will still work fine even if the assert fails.  I should be up to `buildPipelineWithRelocatableElf` to make sure it did not do anything unexpected.  It is not the responsibility of each of the users of the function.

File: llpc/include/llpc.h
Comment (jaebaek) [reply llpc/include/llpc.h]: done

File: llpc/include/llpc.h
Comment (jaebaek) [reply llpc/include/llpc.h]: done

File: llpc/include/llpc.h
Comment (jaebaek) [reply llpc/include/llpc.h]: done

File: llpc/include/llpc.h
Comment (s-perron) [inline llpc/include/llpc.h:113]: I don't know if I like "Invalid" for this name.  It gives the impression that something went wrong.  Something like "CacheUnchecked" might be better.

The comment should be clearer.  From this comment is it not clear why it should be invalid.  You could something like "The cache was not checked."

File: llpc/context/llpcCompiler.cpp
Comment (s-perron) [inline llpc/context/llpcCompiler.cpp:779]: You did not add a comment for the new parameter.  Be sure to call out that you expect it to already be initialized.

File: llpc/context/llpcCompiler.cpp
Comment (s-perron) [inline llpc/context/llpcCompiler.cpp:1433]: Need to add a comment for the new parameter.

File: llpc/include/llpc.h
Comment (jaebaek) [reply llpc/include/llpc.h]: done

File: llpc/context/llpcCompiler.cpp
Comment (s-perron) [inline llpc/context/llpcCompiler.cpp:1439]: I don't see the point of this assert as I mentioned in another comment.

**Review by jaebaek — COMMENTED** [review]
Thank you for your code review. I updated the code based on it.
In terms of `LLPC_INTERFACE_MAJOR_VERSION`, I used `grep` to understand how I have to use it but I have no idea. Could you please give me some advice how to use it?

**Review by s-perron — CHANGES_REQUESTED** [review]
You should update the PR and commit descriptions to include the reason we are adding this.  It is part of implementing the Vulkan Pipeline creation feedback, and point to a change in XGL that will make use of this to implement that feature.  I think people will want to see the whole thing.

File: llpc/include/llpc.h
Comment (jaebaek) [reply llpc/include/llpc.h]: done.

**Review by jaebaek — COMMENTED** [review]
Thank you for code review. PTAL!

File: llpc/context/llpcCompiler.cpp
Comment (kuhar) [reply llpc/context/llpcCompiler.cpp]: `ArrayRef` is only for referencing array-like continuous containers, e.g., `std::vector`, `llvm::SmallVector`, `std::array`, or plain C arrays of static size. Unlike the last two, `ArrayRef` doesn't have a size parameter in its type.

I'd recommend using `std::array` in the place where the source array is defined, and then passing it by `ArrayRef` or `std::array` reference.

Comment (kuhar) [reply llpc/context/llpcCompiler.cpp]: And `MutableArrayRef` is a mutable version of `ArrayRef` that allows to modify the underlying data.

**Comment by kuhar** [comment]
@jaebaek 
> Thank you for your code review. I updated the code based on it.
> In terms of `LLPC_INTERFACE_MAJOR_VERSION`, I used `grep` to understand how I have to use it but I have no idea. Could you please give me some advice how to use it?

It's defined in `include/vkgcDefs.h`, you can see how it's used to guard some code in https://github.com/GPUOpen-Drivers/llpc/commit/9ce37dd671bfabaf30bcbd12c473e7414d8375f0.

File: llpc/context/llpcCompiler.cpp
Comment (kuhar) [inline llpc/context/llpcCompiler.cpp:64]: undo this change

File: llpc/context/llpcCompiler.cpp
Comment (kuhar) [inline llpc/context/llpcCompiler.cpp:784]: Don't use C arrays as function arguments. Prefer `llvm::ArrayRef`, or `llvm::MutableArrayRef` when the size is static but not known, `std::array<...>` when the size is static and known, or `llvm::SmallVectorImpl` when the size is dynamic.

File: llpc/context/llpcCompiler.cpp
Comment (kuhar) [inline llpc/context/llpcCompiler.cpp:1439]: Same as above

File: llpc/include/llpc.h
Comment (kuhar) [inline llpc/include/llpc.h:121]: You use 3 values to describe stage cache accesses (`ShaderCacheInfo`) and only 2 values to describe pipeline cache access (hit/miss). Does the code always check pipeline cache, or does the vulkan API force us to return true/false here?

**Review by kuhar — CHANGES_REQUESTED** [review]

File: include/vkgcDefs.h
Comment (jaebaek) [inline include/vkgcDefs.h:46]: @kuhar Does this version update need a special approval from AMD?

**Review by jaebaek — COMMENTED** [review]
I squashed the commits and updated `LLPC_INTERFACE_MAJOR_VERSION`.

File: include/vkgcDefs.h
Comment (kuhar) [reply include/vkgcDefs.h:46]: We should wait for a review from AMD to confirm.

File: llpc/context/llpcCompiler.cpp
Comment (kuhar) [inline llpc/context/llpcCompiler.cpp:833]: You probably want to assert somewhere near the top of the loop that `stageCacheAccess` has enough elements to avoid potential OOB errors if the stages change in the future.

File: llpc/context/llpcCompiler.h
Comment (kuhar) [inline llpc/context/llpcCompiler.h:130]: Consider taking `stageCacheAccess` by reference

**Review by kuhar — COMMENTED** [review]
Looks fine to me, thanks for the fixes. Please wait for reviews from Steven and AMD.

**Review by kuhar — APPROVED** [review]

File: llpc/context/llpcCompiler.cpp
Comment (jaebaek) [reply llpc/context/llpcCompiler.cpp:834]: done

File: llpc/context/llpcCompiler.h
Comment (jaebaek) [reply llpc/context/llpcCompiler.h:130]: `ElfPackage *pipelineElf` uses a pointer for the output data.
Could you please explain why `stageCacheAccess` should be a reference?

File: include/vkgcDefs.h
Comment (amdrexu) [inline include/vkgcDefs.h:73]: Since you just add new members to structure info, I think a minor version bump is enough 45.1. If we remove some existing or change the interface method, we need to do a major version bump.

**Review by amdrexu — COMMENTED** [review]
LGTM. Not sure if Steven or Tim will take a look.

File: llpc/context/llpcCompiler.h
Comment (kuhar) [reply llpc/context/llpcCompiler.h:130]: It makes it clear that there's a single non-null object behind it and there's no pointer arithmetic.

File: llpc/context/llpcCompiler.cpp
Comment (s-perron) [inline llpc/context/llpcCompiler.cpp:780]: I don't know the technical definitions, but, if you expect that that the parameter is initialized, should it be an in/out parameter?  On the other hand, that initial value is never used, it is just so the caller can differentiate between the values that were set and those that were not.

File: llpc/context/llpcCompiler.cpp
Comment (s-perron) [inline llpc/context/llpcCompiler.cpp:1526]: I think this might be in the wrong place.  The `lookUpShaderCaches` is also a cache lookup, and should set the `pipelineCacheAccess`.  Can you move setting this value to include the lookup in the other cache?

File: llpc/context/llpcCompiler.cpp
Comment (s-perron) [inline llpc/context/llpcCompiler.cpp:1665]: Same as above.

**Review by s-perron — APPROVED** [review]
I just have a couple minor comment.  I'm fine even if they do not get fixed.

**Review by amdrexu — APPROVED** [review]

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit b8edb5ede8d719c49f6c3c19431cc4b1195628c6
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev eb28c7ef53329f838e693f233696823330e964c0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 63761ecfd2d3cfa61a756ce0a72e1ca1cdae333a</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev fde3d4e2bd90743606ca7b724bf62a53cacf4615</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/144351)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26321/48174 (54.6%)</li>
<li>Failed: 0/48174 (0.0%)</li>
<li>Skipped: 21853/48174 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26320/48174 (54.6%)</li>
<li>Failed: 0/48174 (0.0%)</li>
<li>Skipped: 21854/48174 (45.4%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Timeout or crash</li>

</ul></details>

<details open><summary>Rhel 8.2, Gfx103</summary><ul>

<li>Passed: 26219/48003 (54.6%)</li>
<li>Failed: 0/48003 (0.0%)</li>
<li>Skipped: 21784/48003 (45.4%)</li>
</ul></details>


</ul></details>

**Pull request merged**
